src/Serilog.Extensions.Logging/Extensions/Logging/EventIdPropertyCache.cs

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Serilog.Extensions.Logging
+{
+    using Microsoft.Extensions.Logging;
+    using Serilog.Events;
+
+    internal sealed class EventIdPropertyCache
+    {
+        private readonly object _createLock = new();
+        private readonly int _maxCapacity;
+        private readonly Dictionary<EventKey, LogEventProperty> _propertyCache;
+
+        private int count;
+
+        public EventIdPropertyCache(int maxCapacity)
+        {
+            _maxCapacity = maxCapacity;
+            _propertyCache = new Dictionary<EventKey, LogEventProperty>(capacity: maxCapacity);
+        }
+
+        public LogEventProperty GetOrCreateProperty(in EventId eventId)
+        {
+            var eventKey = new EventKey(eventId);
+
+            if (_propertyCache.TryGetValue(eventKey, out var cachedProperty))
+            {
+                return cachedProperty;
+            }
+
+            lock (_createLock)
+            {
+                // Double check under lock
+                if (_propertyCache.TryGetValue(eventKey, out cachedProperty))
+                {
+                    return cachedProperty;
+                }
+
+                cachedProperty = CreateCore(in eventKey);
+
+                if (count < _maxCapacity)
+                {
+                    _propertyCache[eventKey] = cachedProperty;
+                    count++;
+                }
+
+                return cachedProperty;
+            }
+        }
+
+        private static LogEventProperty CreateCore(in EventKey eventKey)
+        {
+            var properties = new List<LogEventProperty>(2);
+
+            if (eventKey.Id != 0)
+            {
+                properties.Add(new LogEventProperty("Id", new ScalarValue(eventKey.Id)));
+            }
+
+            if (eventKey.Name != null)
+            {
+                properties.Add(new LogEventProperty("Name", new ScalarValue(eventKey.Name)));
+            }
+
+            return new LogEventProperty("EventId", new StructureValue(properties));
+        }
+
+        private readonly struct EventKey : IEquatable<EventKey>
```

Comment:
`readonly record struct EventKey(...)` would be sufficient (compiler provides equality and hash code).